### DISCUSSION TITLE
[gatsby-source-contentful] Some referenced nodes can't be resolved during onCreateNode lifecycle

### DISCUSSION THREAD

**Discussion by osartun** [gatsbyjs/gatsby#32414] [Help]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

We have a local transformer plugin built on top of `gatsby-source-contentful` which breaks when we update the Contentful plugin. The problem is that `getNode(…)` returns `null` for certain referenced nodes even though the nodes exist at the end of the bootstrap process.

The issue didn't exist until `2.1.100`. It must have been introduced with https://github.com/gatsbyjs/gatsby/pull/22307 when sourcing nodes was turned asynchronous.

Related links:
* I wrote about my findings in a comment on this issue: https://github.com/gatsbyjs/gatsby/issues/22792#issuecomment-874037765
* It's possible that this is the same issue: https://github.com/gatsbyjs/gatsby/issues/32290

### Reproduction Link

https://github.com/osartun/gatsby-contentful-transformer-issue

### Steps to Reproduce

Clone the reproduction repo and run gatsby in the `v3` folder. Read more in the README of the reproduction repo.

Alternatively:

1. Use `gatsby-source-contentful >= 2.2.0` 
2. Have a Contentful entry reference another entry of a different content type. (It might be important that the referenced content type was created after your entry)
3. In your `gatsby-node.js` in the `onCreateNode` lifecycle call `getNode()` with the ID of the referenced node

### Expected Result

`getNode({referenced id})` returns the node

### Actual Result

`getNode({referenced id})` returns null

### Environment

```shell
System:
    OS: macOS 11.3.1
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.17.1 - ~/.nvm/versions/node/v14.17.1/bin/node
    Yarn: 1.22.10 - ~/.nvm/versions/node/v14.17.1/bin/yarn
    npm: 6.14.13 - ~/.nvm/versions/node/v14.17.1/bin/npm
  Languages:
    Python: 2.7.16 - /usr/bin/python
  Browsers:
    Chrome: 91.0.4472.114
    Firefox: 89.0.1
    Safari: 14.1
  npmPackages:
    gatsby: ^3.9.1 => 3.9.1
    gatsby-plugin-image: ^1.9.0 => 1.9.0
    gatsby-source-contentful: ^5.9.0 => 5.9.0
  npmGlobalPackages:
    gatsby-cli: 3.9.0
```

**Comment by axe312ger** [comment]
Thank you for the detailed report. I'll have a look. @wardpeet Seems like your changed introduced the issue, any idea why?

**Comment by wardpeet** [comment]
My guess is that this was luck that it worked before. The order how we create nodes probably got changed. It's not a good idea to fetch other nodes during onCreateNode call. @osartun I suggest creating a customer resolver instead of adding it as a field to the object.

**Comment by osartun** [comment]
@wardpeet You mean the `createResolvers` API? https://www.gatsbyjs.com/docs/reference/graphql-data-layer/schema-customization/#createresolvers-api

**Comment by wardpeet** [comment]
@osartun https://www.gatsbyjs.com/docs/reference/graphql-data-layer/schema-customization/

this comes closest to what you're looking for
https://www.gatsbyjs.com/docs/reference/graphql-data-layer/schema-customization/#foreign-key-fields
